File: dsa-courses/colt-steele/12-stacks-and-queues/stack.js
class Node {
    constructor(val) {
        this.val = val
        this.next = null
    }
}

class Stack {
    constructor() {
        this.first = null
        this.last = null
        this.size = 0
    }

    push(val) {
        let newNode = new Node(val)
        if (this.size == 0) {
            this.first = newNode
            this.last = newNode
        } else {
            let firstElem = this.first
            this.first = newNode
            newNode.next = firstElem
        }
        return ++this.size
    }

    pop() {
        if (!this.first) {
            return null
        }
        let temp = this.first
        if (this.size === 1) {
            this.first = null
            this.last = null
        } else {
            this.first = this.first.next
            temp.next = null
        }
        this.length--
        return temp.val
    }
}

let stack = new Stack()
console.log(stack.push("FIRST"))
console.log(stack.push("SECOND"))
console.log(stack.push("THIRD"))
console.log(stack.push("LAST"))
console.log(stack.pop())

File: dsa-courses/python-course/python-dsa-revision/stacks-and-queue/stack.py
class Node:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = None

class Stack:
    def __init__(self):
        self.first = None
        self.last = None
        self.size = 0

    # add to the "head" of the linked list because this is a stack
    def push(self, val):
        newNode = Node(val)

        # no nodes
        if self.size == 0:
            self.first = newNode
            self.last = newNode
        else:
            firstElement = self.first
            self.first = newNode
            newNode.next = firstElement

        self.size += 1
        return self.size
    
    # remove the "head" of the linked list
    def pop(self):
        if self.first is None:
            return None
        
        removedNode = self.first
        if self.first == self.last:
            self.first = None
            self.last = None
        else:
            self.first = self.first.next
            removedNode.next = None

        self.size -= 1
        return removedNode.val
    
S = Stack()
print(S.push("First"))
print(S.push("Second"))
print(S.push("Third"))
print(S.push("Fourth"))
print(S.pop())